### ISSUE TITLE
the permissions are too broad/scary

### ISSUE THREAD

**Issue by wmhilton** [all-contributors/all-contributors-bot#252]
I can't bring myself to grant a bot "Write access to code" permission. Is there _any_ way to work around this?

![image](https://user-images.githubusercontent.com/587740/67174313-96f4a880-f38f-11e9-930d-3101f9a7e218.png)

**Comment by gr2m** [comment]
Unfortunately there is not at this point. The app needs to update more than one file, so the single file permission would not suffice.

The app cannot however write to your default branch if you have branch protection enabled.

**Comment by laurensV** [comment]
Also, with the single file permission you can request access to as many files as you want, so there is no reason to ask write access to all code. You can understand that legit open-source projects that take security seriously would and should never give that permission to an unknown third party

**Comment by laurensV** [comment]
> Not sure what you are trying to say. But it sure reads unnecessarily confrontational. Please be kind, this is a community run open source project, you make assumptions with very little context.

You are right, it came from my assumption that you would only need access to 1 README file, but I guess its more complex than that. My apologies, appreciate these kind of initiatives!

**Comment by tenshiAMD** [comment]
@laurensV BTW, if you are into using the bot, you can use the CLI instead. https://github.com/all-contributors/app. I think we can close this one. @wmhilton @laurensV do you still have any other concerns?

**Comment by laurensV** [comment]
No other concerns except for the security risk already mentioned which I still think is a major concern. I would still propose/prefer a solution where instead of asking write permission for all code (so that users can specify any files that they want to write to themselves in the `files` attribute of the `.all-contributorsrc` file), single file permission is asked for a certain set of files where people can choose from (I've checked repositories using this file and it seems like 99% are using `README.md` or `docs/README.md`). Also tried to find info on the limit of files you can ask permission for in single file permission but could not find the limit, it seems like you can add at least hundred files if you want there. Where could I find the limit? The list of files that this app could ask write permission for could look something like this (this would cover more than 99% of all repos):
```
README.md
README
readme.md
readme
docs/README.md
docs/README
docs/readme.md
docs/readme
AUTHORS.md
AUTHORS
authors.md
authors
docs/AUTHORS.md
docs/AUTHORS
docs/authors.md
docs/authors
CONTRIBUTING.md
CONTRIBUTING
contributing
contributing.md
docs/CONTRIBUTING.md
docs/CONTRIBUTING
docs/contributing
docs/contributing.md
```

It does restrict the number of files you can choose to write to, but it comes with greatly improved security benefits for all repos that use this app. The ones that really want to write to custom files could/should fall back to the cli, while good security would then be the standard for all repos.

Now 100% of all repos that are using this app have a security risk for 1% of the repos that want to write to custom files..

If you decide to not sacrifice the flexibility of being able to write to any custom file and keep this security risk, this ticket can be closed imo.

PS. Maybe I am misunderstanding or missing something for the need for the write permission for all code. Please let me know if that is the case

**Comment by tenshiAMD** [comment]
@laurensV allow me to summarize what you said. Are you saying to remove the `files` config option and just statically define certain file paths? I guess that could be a workaround. @gr2m what do you think?

**Comment by laurensV** [comment]
@tenshiAMD Yeah almost: I propose to remove the 'write all code' permission and instead ask single file permission for a predefined file paths (e.g. the list in my previous post). Then people can still use the `files` config option, but only file paths from the predefined list are allowed. So you lose the flexibility to define any filepath you want in `fiiles`, but it removes the security risk of having a third party be able to write all your code (especially for projects that use ci/cd with autodeploy enabled this can be very dangerous)

**Comment by tenshiAMD** [comment]
>  'write all code' permission

@laurensV I think you got it wrong.

> a third party be able to write all your code (especially for projects that use ci/cd with autodeploy enabled this can be very dangerous)

@laurensV I think you just misunderstood something in the implementation. The app is ONLY allowed to write are those defined in the `files` config option, therefore, it means the end users are fully aware that they will grant write access ONLY to those files for the bot to process; which make sense that's why there is an option for that.

**Comment by laurensV** [comment]
> @laurensV I think you got it wrong. It never stated "ALL".
@tenshiAMD If you look at the screenshot from the original post, it clearly asks for the `Write access to code` permission, which is a permission that allows this app to change ALL code, so I don't think I got that wrong..

> I think you just misunderstood something in the implementation. The app is ONLY allowed to write are those defined in the files config option
That is not true. the app in theory is allowed to write ALL files (hence the permission), but is now only writing to files defined in the `files` config, but nothing at the moment could stop the app to also write other files (meaning if this single app gets hacked, the hacker will get access to write/change all the code of all repositories that are using this app. To make sure the app is never allowed to do this, you need to change the `Write access to code` permission to the single file permissions for the github app

**Comment by gr2m** [comment]
> which is a permission that allows this app to change ALL code

not if you set branch protections, which you should do anyway
